Idling and Unidling

DirectXMan12 opened the pull request:
This PR introduces idling and unidling.

Idling works by scaling scalable resources currently (DCs and RCs) down to zero, and then marking endpoints objects for associated services to indicate that traffic on those services should unidle those scalable resources.

The `oc idle` command does the actual work of associating the scalables with services (using the graph API), annotating the endpoints, and scaling the services down to zero.  The command can be passed a file containing either a list of services or a list of scalables (all within the current namespace).

Unidling works by creating a new kube-proxy `Proxy` implentation, the `HybridProxy`.  The `HybridProxy` listens like a normal proxy, and forks operations out to a "normal" proxy or a special unidling proxy.  The unidling proxy listens for TCP connections, holds those connections, and emits a `NeedPods` event (see below), passing the traffic through once the pods come up.  UDP packets trigger the same `NeedPods` event, but are subsequently dropped on the floor.

The unidling controller then listens for these events (see below), and "wakes up" the appropriate scalables by scaling the back to a replica count of 1.

*Below*: An alternative to using events would be to use an extra annotation on the endpoints object to trigger unidling (using annotations seems slightly more in line with how Kube works than events, IMO).  Various parties have expression opinions about whether or not events or some other mechanism should be used to trigger the unidling.  I wanted to get some feedback before going an rewriting part to use another mechanism.

Once decisions are made about the annotation names, etc, this PR should be sufficient to start work on necessary UI components.

For the next PR?:
- [ ] Collecting candidates for idling based on metrics (this is waiting on reviews, etc for Oldtimer in upstream Heapster)
- [ ] Recording network traffic on a service-level to help make better idling decisions (this is waiting on reviews, etc for Push Metrics in upstream Heapster)
- [ ] Not using events to trigger unidling

TODO/Complete:
- [x] Idling a provided list of services or scalables
- [x] Unidling (using events)
- [x] unit tests for the unidling controller
- [x] retrying controller
- [x] record scale and scale back to full instead of 1

Thanks to @knobunc for assistance with the implementation and for the in-progress work on network metrics collection.

DirectXMan12 commented:
cc @smarterclayton @ncdc @knobunc @pmorie @derekwaynecarr @liggitt @abhgupta

smarterclayton commented:
Is the HybridProxy using a specific implementation, or does it work with both iptables and userspace?

DirectXMan12 commented:
The "normal proxy" part of the hybrid proxy can be anything implementing the normal proxy interface, so both iptables and userspace should work there (most of my personal testing was with iptables, though).  The "unidling" part of the hybrid proxy needs to be userspace, but as soon as the pods wake up, the hybrid proxy switches incoming connections back to the "normal" proxy.

smarterclayton commented:
Ok, will look through this more tonight.  Thanks

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/hybrid/proxier.go:
```diff
@@ -0,0 +1,130 @@
+package hybrid
```
Probably shouldn't be upstream for now, unless you're ready to open the PR to Kube.

DirectXMan12 replied:
I can move it into a subpackage of idling in origin, perhaps.

smarterclayton replied:
We should come up with a clever name that means both idling and unidling, and then use "unidle" and "idle" as our package names ("ing" is untidy).  We can worry about that right at the end though, but people should suggest good names that cover both parts.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/roundrobin.go:
```diff
@@ -112,6 +112,17 @@ func isSessionAffinity(affinity *affinityPolicy) bool {
+	defer lb.lock.Unlock()
+	if state, exists := lb.services[svcPort]; exists && state != nil && len(state.endpoints) > 0 {
+		return true
+	} else {
```
style, don't have an else when you return.

smarterclayton replied:
and also this shouldn't be an if - `state, exists := ....; return state != nil && ...`.

DirectXMan12 replied:
yep

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/proxier.go:
```diff
@@ -413,7 +435,12 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
 
 			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, serviceIP, servicePort.Port, servicePort.Protocol)
-			info, err = proxier.addServiceOnPort(serviceName, servicePort.Protocol, proxyPort, proxier.udpIdleTimeout)
+			svcRef, err := api.GetReference(service)
```
GetReference is a really bad idea.  I'd just create an Object reference and pass it in to addServiceOnPort (as a value, not a pointer) and remove `serviceName`.  Set Kind, Namespace, Name, and UID.  Don't set anything else (unless you need it).

DirectXMan12 replied:
Is the full reference not needed for the event recorder interface?

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+/*
```
This seems like it belongs in a different package.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+		ticker := time.NewTicker(needPodsTickLen)
+		allConns = newConnectionList(maxHeldConnections, needPodsTickLen, needPodsWaitTimeout, svcName)
+
+	WAIT_LOOP:
```
Go style is use `WaitLoop:`.

smarterclayton replied:
Also, if you need to use this, you should probably refactor to call a function.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/loadbalancer.go:
```diff
@@ -30,4 +30,5 @@ type LoadBalancer interface {
 	NextEndpoint(service proxy.ServicePortName, srcAddr net.Addr, sessionAffinityReset bool) (string, error)
```
I want to believe this could be upstreamed as is, if you can cleanly decouple it and leave comments throughout the code about it.

Do you have an idling proposal that reflects this code that upstream has even seen?

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/proxier.go:
```diff
@@ -140,6 +145,21 @@ func IsProxyLocked(err error) bool {
+
+// NewUnidlerProxier works just like NewProxier, except that it creates an unidling proxy
+// designed to work as part of the hybrid proxier.
+func NewUnidlerProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables.Interface, pr utilnet.PortRange, syncPeriod, udpIdleTimeout time.Duration, recorder record.EventRecorder) (*Proxier, error) {
```
Would prefer this is in a different package.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+
+func (udp *udpUnidlerSocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
+	// just drop the packets on the floor until we have endpoints
+	var buffer [4096]byte // 4KiB should be enough for most whole-packets
```
All these magic numbers should be package private constants.

DirectXMan12 replied:
Fair, that line was just copied from the proxysocket.go version

smarterclayton replied:
Yeah proxy is not always great code (it's some of the surviving code from the very first kube days - we had to write our own functions in those days, and merge commits uphill both ways in the snow)

DirectXMan12 replied:
![](http://i.imgur.com/9rYVxqH.jpg)

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+// no new connections are allowed but existing connections are left untouched.
+type tcpUnidlerSocket struct {
+	tcpProxySocket
+	eventRecorder record.EventRecorder
```
If you don't need all the methods on EventRecorder, make this be a simpler interface (with whichever method you need).

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+				}
+
+				if !sent_need_pods && !proxier.loadBalancer.ServiceHasEndpoints(service) {
+					// TODO: There's a constant in origin for this NeedPods
```
Don't define cross package constants.  Define an internal constant here, and use another elsewhere.  Constants that cross large package boundaries are false coupling.

DirectXMan12 replied:
Ack.  I've moved most of the stuff into a package in origin, so now I just have a NeedPods constant at the top of this file (which may get moved when I abstract the signal function out anyway).

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+					// TODO: There's a constant in origin for this NeedPods
+					// HACK: make the message different to prevent event aggregation
+					glog.V(4).Infof("unidling TCP proxy sent unidle event to wake up service %s/%s:%s", service.Namespace, service.Name, service.Port)
+					tcp.eventRecorder.Eventf(myInfo.ref, api.EventTypeNormal, unidling.NeedPodsReason, "The service-port %s/%s:%s needs pods (at %v)!", service.Namespace, service.Name, service.Port, time.Now())
```
This should be abstracted into a "Signal" function, rather than being an explicit event.  Take as arguments your inputs.

DirectXMan12 replied:
Yep.  That also should allow us to slot in a different signal function easily depending on what we decide.

derekwaynecarr replied:
is the namespace needed in the message?  is it not inferred from the namespace of the event itself?

derekwaynecarr replied:
why is ports enumerated?  isn't the idea that a service has no endpoints and therefore needs pods (independent of a port)?

DirectXMan12 replied:
> is the namespace needed in the message? is it not inferred from the namespace of the event itself?

Yep, removed

> why is ports enumerated? isn't the idea that a service has no endpoints and therefore needs pods (independent of a port)?

Sure, but independently, it could be useful to see that information in the event (If I'm a user or admin looking at what caused something to unidle, etc)

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+	}
+}
+
+func (tcp *tcpUnidlerSocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
```
Don't use `my*` prefix for variables, call it `serviceInfo`.

smarterclayton replied:
Also, `servicePortName` or just `name` instead of `service`.

DirectXMan12 replied:
agreed (was a holdover from the normal proxy socket code ;-) )

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+				return true
+			}
+		}
+		glog.Errorf("ReadFrom failed, exiting ProxyLoop: %v", err)
```
You're not exiting proxy loop in this code path.  Is this really important?

DirectXMan12 replied:
I've added some doc to this function, but basically returning false from this function means "don't try to read again from this, and exit the proxy loop instead".

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+	// just drop the packets on the floor until we have endpoints
+	var buffer [4096]byte // 4KiB should be enough for most whole-packets
+
+	glog.V(4).Infof("unidling proxy UDP proxy waiting for data")
```
include which actual service and port you're unidling

DirectXMan12 replied:
ack

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+
+	for {
+		if !udp.readFromSock(buffer[0:], myInfo) {
+			break
```
Your log message should be here.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+		return
+	}
+
+	stopChan, wakeupTimeoutTimer := udp.sendWakeup(service, myInfo)
```
If I UDP flood you, can I cause the proxy to create essentially infinite numbers of these timers and stop chans and goroutines?  If not, comment why not in the godoc for this method.

smarterclayton commented on pkg/api/kubegraph/nodes/types.go:
```diff
@@ -69,10 +69,6 @@ func (n EndpointsNode) String() string {
 	return string(EndpointsNodeName(n.Endpoints))
 }
 
-func (n EndpointsNode) ResourceString() string {
```
?

smarterclayton commented on pkg/cmd/server/origin/master_config.go:
```diff
@@ -641,6 +641,12 @@ func (c *MasterConfig) OriginNamespaceControllerClients() (*osclient.Client, *kc
 
+// UnidlingControllerClients returns the unidling controller clients
+func (c *MasterConfig) UnidlingControllerClients() (*osclient.Client, *kclient.Client) {
+	// TODO(directxman12): make this use service account clients
```
Will be required for merge - follow the pattern for the other ones (need to create role and SA).

smarterclayton commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -503,3 +505,13 @@ func (c *MasterConfig) RunResourceQuotaManager(cm *cmapp.CMServer) {
+// RunUnidlingController starts the unidling controller
+func (c *MasterConfig) RunUnidlingController() {
+	oc, kc := c.UnidlingControllerClients()
+	resyncPeriod := 2 * time.Minute
```
Does this need to be this short?

DirectXMan12 replied:
No, I don't think so.  Do we have a recommended time interval for resyncs written down somewhere.

smarterclayton replied:
Basically at this point unless you need to check external state (and given the comments about listing old events being not that useful) you can set this to hours.

smarterclayton commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -503,3 +505,13 @@ func (c *MasterConfig) RunResourceQuotaManager(cm *cmapp.CMServer) {
+func (c *MasterConfig) RunUnidlingController() {
+	oc, kc := c.UnidlingControllerClients()
+	resyncPeriod := 2 * time.Minute
+	delegScaleNamespacer := osclient.NewDelegatingScaleNamespacer(oc, kc)
```
Just call this `scaleNamespacer`

DirectXMan12 replied:
ack

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+package unidling
```
`pkg/service/controller/unidling`

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+
+	_, controller := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
```
I think you can just avoid relist and resync altogether.  If we miss an unidle it's not the end of the world.

DirectXMan12 replied:
fair.  Just have an empty list function?

smarterclayton replied:
Yeah.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/unidling"
```
Don't put this in Kube.  Also, it's not util.

DirectXMan12 replied:
yep, I've moved it.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+			continue
+		}
+
+		scale.Spec.Replicas = 1
```
I think this is potentially wrong (and possibly very wrong).  At the time we idle we need to record the historical scale, and unidle needs to respect that.

DirectXMan12 replied:
I think we discussed this a bit in the unidling doc, but, IMO, it doesn't make sense to restore to "full" because the scalable has presumably been idled since it hasn't received any traffic (or was very low on some other metric).

If I have 5 replicas, but I've received no traffic in the past 24h, chances are I'm probably not going to see a lot of traffic after wakeup (although I could see a scenario where somebody is aggressively idling, and this is not the case).

If we think waking up to "full" is a thing people are going to want, then it should probably be an option to oc idle, with a default to "wake up to 1".

What do other people think?

ncdc replied:
I can see utility in going from 0 to 1, but I believe we also talked about the need to unidle a cluster of pods, where replicas=1 doesn't make sense or work when you're unidling. Does that sound familiar?

DirectXMan12 replied:
yeah, I talked to @mpbarrett earlier and he brought up that you'd probably want to always have some minimum number of pods around, so I see the utility there.  I'd probably still add a flag to allow you to choose, but I see the point.

smarterclayton replied:
Example I had was a petset.  It requires 2 to function, but can be idled.  You can't just go back to 1 (it wouldn't work anymore), so we have to go back to the previous scale value.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+		scale.Spec.Replicas = 1
+
+		if _, err := scales.Update(scalableRef.Kind, scale); err != nil {
+			glog.Errorf("Unable to scale up %s %q while unidling service %s/%s: %v", scalableRef.Kind, scalableRef.Name, targetNamespace, targetServiceName, err)
```
Any Errorf in here should be util.HandleError().

DirectXMan12 replied:
ack.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+		newAnnotationList = append(newAnnotationList, k)
+	}
+
+	// TODO(directxman12): do we need to re-fetch the service to avoid conflicts while updating?
```
Comment is wrong.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+				event := obj.(*kapi.Event)
+
+				if event.Reason != unidling.NeedPodsReason {
+					glog.Errorf("UnidlingController was asked to handle a non-NeedPods event")
```
Not much point to this message, but is either util.HandleError or just V(4).

DirectXMan12 replied:
Yeah, I think that snuck in from debugging.  Good catch.

smarterclayton commented on pkg/unidling/controller.go:
```diff
@@ -0,0 +1,178 @@
+					return
+				}
+
+				if err := unidlingController.handleEvent(event); err != nil {
```
We need to make a decision whether you are a retrying queue.  I think you have to be - one try isn't enough.

DirectXMan12 replied:
Seems like a good idea.

sttts commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/hybrid/proxier.go:
```diff
@@ -0,0 +1,130 @@
+
+	// TODO(directxman12): figure out a good way to avoid duplicating this information
+	// (it's saved in the individual proxies as well)
+	usingUserspace map[types.NamespacedName]bool
```
style nit: `map[types.NamespacedName]struct{}`. You use the `_, found := x[key]` lookup style anyway.

DirectXMan12 replied:
yeah, I'm using it the same way (I was just using the `bool` type instead of `struct`).  I'll change it

sttts commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/hybrid/proxier.go:
```diff
@@ -0,0 +1,130 @@
+	p.unidlingLoadBalancer.OnEndpointsUpdate(endpoints)
+	p.mainLoadBalancer.OnEndpointsUpdate(forIPTables)
+
+	p.OnServiceUpdate(p.serviceConfig.Config())
```
Why do we have to cross-call `OnServiceUpdate` here. Probably worth a comment.

sttts commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/proxier.go:
```diff
@@ -140,6 +145,21 @@ func IsProxyLocked(err error) bool {
+
+// NewCustomProxier works just like NewProxier, except that it takes a custom function for actually
+// proxying the socket.
+func NewCustomProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables.Interface, pr utilnet.PortRange, syncPeriod, udpIdleTimeout time.Duration, makeProxySocket ProxySocketFunc) (*Proxier, error) {
```
does this have to be public?

sttts commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+	return nil, fmt.Errorf("unknown protocol %q", protocol)
+}
+
+// tcpUnidlerSocket implements proxySocket.  Close() is implemented by net.Listener.  When Close() is called,
```
nit: some spaces too much

DirectXMan12 replied:
??

derekwaynecarr commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go:
```diff
@@ -0,0 +1,365 @@
+			return
+		case <-time.NewTicker(needPodsWaitTimeout).C:
+			glog.V(4).Infof("unidling proxy sent unidle event to wake up service %s/%s:%s", service.Namespace, service.Name, service.Port)
+			udp.eventRecorder.Eventf(myInfo.ref, unidling.NeedPodsReason, "The service-port %s/%s:%s needs pods (at %v)!", service.Namespace, service.Name, service.Port, time.Now())
```
this is the same basic event as the one earlier in this pull.  should there be a shared func so they match?

smarterclayton commented:
One wrinkle with events - we want to make sure unidle events are compressed (so only the number changes) which means the messages need to be stable. 

We may also want a global rate limiter in the controller on how fast things get unidled (once we have workqueues).

DirectXMan12 commented:
Ok, I've think I've gotten all of the comments, except for the retrying controller.  For that, is there a good example of one of those?  I saw a couple of slightly different approaches (kube's service controller, for example), and was wondering if there was a canonical example/existing component that I should be using.

DirectXMan12 commented:
also, in order to move stuff around, I ended up having to expose a some things from the userspace proxy.   It's not perfect, but it works.

DirectXMan12 commented:
Also need to do the scale back to full code too

smarterclayton commented on pkg/router/controller/factory/factory.go:
```diff
@@ -25,30 +25,32 @@ import (
-	Namespace      string
-	Labels         labels.Selector
-	Fields         fields.Selector
+	EndpointsNamespacer kclient.EndpointsNamespacer
```
Is there a reason we aren't writing the annotation to the endpoints object instead of the service?

smarterclayton replied:
Wouldn't the endpoints be the natural place to look?

smarterclayton replied:
In the long run we have to load services (for ingress).  In the short run I'd really prefer not to.  If we have the endpoints controller copy the idle annotation to endpoints, would that be sufficient to know here?

DirectXMan12 replied:
> Is there a reason we aren't writing the annotation to the endpoints object instead of the service?

We do write the annotation to the endpoints object...

> would that be sufficient to know here?

No, you have to know the service IP so that the router can target requests to the unidling proxy (instead of just returning the "no endpoints" HTTP error) for idled services.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,539 @@
+// scanIdlableResourcesFromFile scans a file for either a list of resource names (when a default resource is provided),
+// or for a list of resources as resource/name (when no default resource is provided).  It returns a map from
+// resolved resource group-kind to a set of resource names.
+func scanIdlableResourcesFromFile(filename string, f *clientcmd.Factory, defaultResource string) (map[unversioned.GroupKind]map[string]bool, error) {
```
This seems relatively complex.  The UI implementation is going to be really rough.  Who is intended to call this?  Admins?

It seems like it would be easier to implement this by starting with the endpoints list:

1. Get all endpoints in the system that represent things that should be idled.
2. For each endpoint:
  1. if it has zero endpoints do nothing
  2. look for a pod in the endpoint list (set by the endpoint controller as target)
  3. load the pod, and see what the controller reference is (owning controller)
  4. if no controller reference, print information to the admin / put it in a different list and continue
  5. add the controller reference to the list of "things to be scaled down" along with a reference to the owning service
3. Dedup all "things to be scaled down"
4. For each "thing to be scaled down"
  1. Set the service annotation (or the endpoint annotation as discussed elsewhere)
  2. Set the thing to be scaled down original annotation
  3. Scale it to zero

That makes this command basically O(Endpoints + Controllers), and is completely generic (doesn't require the graph library).  The inputs are either "all endpoints", or "list of endpoints".

DirectXMan12 replied:
>  The UI implementation is going to be really rough. Who is intended to call this? Admins?

Yeah, as far as I knew, the intended audience for the idling command was admins.  Is this getting a full equivalent in he web UI?  I was under the impression that since this was an admin command, mainly, that would not be the case.

> It seems like it would be easier to implement this by starting with the endpoints list

So, there is a mode which loads from endpoints instead, but it also currently uses the graph library (you can either start with a list of RCs and DCs to idle and find the services, or have a list of services to idle and find the DCs and RCs).

> That makes this command basically O(Endpoints + Controllers), and is completely generic (doesn't require the graph library). The inputs are either "all endpoints", or "list of endpoints".

So, it seems like that list of steps is kind of rewriting something similar to graph library, but without using the graph library.  What do we gain by dropping the graph library here?  Is the concern not being able to upstream as easily?  Is it not requiring extra calls (I already restrict the graph library to only load services, DCs, RCs, and endpoints, so it's not like we're loading the entire world here currently)?

smarterclayton replied:
I don't think the graph library is relevant here.  You want to fetch endpoints, then the pod (directly) then look up the controller and make a generic "fetch this call", with the appropriate caching / flagging for "i've already loaded this".  Not really a graph use case - the graph assembles large chunks of data at once, this is more iterative.

DirectXMan12 commented:
Ok, I've added a retrying workqueue controller (although we only retry in certain situations), modified the `oc idle` command to work based off of references instead of the graph library, and moved the code required for the unidler socket entirely into its own package (so there's now a fork of the userspace proxy in Origin, as per discussions with @smarterclayton).  We also now record scale and restore to previous scale.

I'd still like to add a couple more unit tests in a couple places (e.g. make sure we're retrying in the correct place, and test a some of the helper functions in the `oc idle` command), and I probably missed a gofmt somewhere.

The only other thing I'm concerned about is that by going based off of references, we refer to the replicationcontroller (or replicaset, if this were to be used with upstream deployments), not the original DC, so we'll be scaling the RCs directly, which is discouraged.  I could add in additional logic that would be specific to DCs (i.e. fetch the RC, check if it has the origin DC label), but it would be nice if there were a more standard annotation we could use to walk up that level (i.e. one that would work with ReplicaSets and Deployments eventually, as well as anything in the future that follows the DC->RC->Pod pattern).  WDYT?

DirectXMan12 commented:
ok, gofmt and govet now pass.  Additionally, I added in some more controller unit tests, and some tests for the idle command helper functions.

The only thing left is the RC->DC logic, any thoughts on that?

smarterclayton commented:
On Jun 28, 2016, at 2:19 PM, Solly Ross <notifications@github.com> wrote:

Ok, I've added a retrying workqueue controller (although we only retry in
certain situations), modified the oc idle command to work based off of
references instead of the graph library, and moved the code required for
the unidler socket entirely into its own package (so there's now a fork of
the userspace proxy in Origin, as per discussions with @smarterclayton
<https://github.com/smarterclayton>). We also now record scale and restore
to previous scale.

I'd still like to add a couple more unit tests in a couple places (e.g.
make sure we're retrying in the correct place, and test a some of the
helper functions in the oc idle command), and I probably missed a gofmt
somewhere.

The only other thing I'm concerned about is that by going based off of
references, we refer to the replicationcontroller (or replicaset, if this
were to be used with upstream deployments), not the original DC, so we'll
be scaling the RCs directly, which is discouraged. I could add in
additional logic that would be specific to DCs (i.e. fetch the RC, check if
it has the origin DC label), but it would be nice if there were a more
standard annotation we could use to walk up that level (i.e. one that would
work with ReplicaSets and Deployments eventually, as well as anything in
the future that follows the DC->RC->Pod pattern). WDYT?


RCs should probably also indicate their owner.  If we walked a few levels
that would be acceptable.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9473#issuecomment-229187013>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_p1am7mS2nhC-aLdLI62amZcxNQ2Gks5qQY_JgaJpZM4I7Lau>
.

smarterclayton commented:
Other than chasing the references, we could add an annotation.  But we
could also make kubectl scale chase references (which is good)

DirectXMan12 commented:
Ok, we now chase both the CreatedByAnnotation from Kubernetes and the DeploymentConfigAnnotation from origin (preferring the former when both a present) up one additional level (enough to go pod->rc->dc).

liggitt commented:
we're idling rc's/dc's based on annotations on pods? I didn't think those annotations were trustworthy

smarterclayton commented:
Does the origin controller not use created by?  If not, please let
@kargakis and @bparees know that they need to be setting those.

On Jun 30, 2016, at 12:36 PM, Solly Ross <notifications@github.com> wrote:

Ok, we now chase both the CreatedByAnnotation from Kubernetes and the
DeploymentConfigAnnotation from origin (preferring the former when both a
present) up one additional level (enough to go pod->rc->dc).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9473#issuecomment-229765498>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_pwfkZRepif_E2oN13YcEx3YMBugLks5qRBq7gaJpZM4I7Lau>
.

smarterclayton commented:
Idling is already not something we can prevent malicious users from
avoiding... They just make connections every X seconds to the server.
Idling isn't security, idling is efficiency.

smarterclayton commented:
The security aspect is the percentage quota stuff - you can only use 18
core hours per day, and we scale you down when you exceed that.

On Jun 30, 2016, at 12:47 PM, Jordan Liggitt <notifications@github.com>
wrote:

we're idling rc's/dc's based on annotations on pods? I didn't think those
annotations were trustworthy

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9473#issuecomment-229768102>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_p49SQVBgNoOoEtmLBIQGvsy8ggmZks5qRB1HgaJpZM4I7Lau>
.

DirectXMan12 commented:
> Does the origin controller not use created by? If not, please let @kargakis and @bparees know that they need to be setting those.

AFAICT, `CreatedByAnnotation` is only set on pods currently.  It would be nice if the Deployment controller and the DC controller would set them as well.

smarterclayton commented:
Let's queue that up

On Jul 1, 2016, at 4:01 PM, Solly Ross <notifications@github.com> wrote:

Does the origin controller not use created by? If not, please let @kargakis
<https://github.com/kargakis> and @bparees <https://github.com/bparees>
know that they need to be setting those.

AFAICT, CreatedByAnnotation is only set on pods currently. It would be nice
if the Deployment controller and the DC controller would set them as well.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9473#issuecomment-230035584>, or mute
the thread
<https://github.com/notifications/unsubscribe/ABG_p2cOhby15_GY7RL6nKEWG39MFrRCks5qRXINgaJpZM4I7Lau>
.

benjaminapetersen commented:
I'm looking at this from the perspective of supporting idled deployments in the web console ([trello](https://trello.com/c/a5rIil2b)).  I'd like to avoid having to open another watch on Endpoints.  If its possible to get the annotation applied to the DCs directly that would definitely be beneficial.

@DirectXMan12 @smarterclayton @jwforres per some conversations we had earlier today.

DirectXMan12 commented:
The issue here is that having the annotation on the DC/RC/etc requires
making the controller aware of the specific objects that it's dealing with
(it needs a general client, etc).  Additionally, it then either has to make
an additional call besides the scale call to update the annotation (which
means another place for things to get out of sync), or has to learn how to
edit the scale of the individual object types (breaking the scale
abstraction and making it less generic).  It's not impossible to add the
annotation, but it makes things more complicated.

On Thu, Jul 7, 2016, 15:34 Ben Petersen <notifications@github.com> wrote:

> I'm looking at this from the perspective of supporting idled deployments
> in the web console (trello <https://trello.com/c/a5rIil2b>). I'd like to
> avoid having to open another watch on Endpoints. If its possible to get the
> annotation applied to the DCs directly that would definitely be beneficial.
>
> @DirectXMan12 <https://github.com/DirectXMan12> @smarterclayton
> <https://github.com/smarterclayton> @jwforres
> <https://github.com/jwforres> per some conversations we had earlier today.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9473#issuecomment-231183523>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHTuczUv-TI_G2vy1SW7r7G4iVhTFZXks5qTVSngaJpZM4I7Lau>
> .
>

smarterclayton commented:
You're setting the annotation onto the object you make the scale call
to, right?  Or not?

kargakis commented on pkg/unidling/controller/controller.go:
```diff
@@ -0,0 +1,323 @@
+	// TODO: ew, this is unversioned.  Such is life when working with annotations.
+	var targetScalables []unidlingapi.RecordedScaleReference
+	if targetScalablesStr, hasTargetScalables := targetEndpoints.Annotations[unidlingapi.UnidleTargetAnnotation]; hasTargetScalables {
+		if err = json.Unmarshal([]byte(targetScalablesStr), &targetScalables); err != nil {
```
You shouldn't be able to unmarshal from an internal object because you shouldn't use json tags in internal objects. Why can't you serialize the versioned representation of the object into the annotation?

DirectXMan12 replied:
I based that somewhat on the code in the HPA controller, which also uses an unversioned type.  Based on a conversation I had with @smarterclayton, I was under the impression that I did not need to introduce an entire new "official" API type, unless I misinterpreted that conversation (which is possible ;-))

smarterclayton replied:
You should have a concrete type that is effectively versioned, but it doesn't have to be round tripped to internal.

deads2k replied:
> You should have a concrete type that is effectively versioned, but it doesn't have to be round tripped to internal.

If you choose to do that, leave a comment to the next guy that if he ever needs another version, he'll need to pay the debt and create the hub, update the handling code to use it, and provide the conversions.

kargakis commented on pkg/unidling/api/types.go:
```diff
@@ -0,0 +1,7 @@
+package api
```
Should be in another commit

DirectXMan12 commented:
> You're setting the annotation onto the object you make the scale call
to, right?  Or not?

Currently, the annotation goes on the endpoints only -- we only interact with the scalable via the scale interface.

kargakis commented:
There are other subresources (see DeploymentConfigRollback[1] in Origin or
DeploymentRollback[2] in Kubernetes) that use an UpdatedAnnotations field
which passes down those annotations to the object it updates. Scale could
probably have something similar, no?

[1]
https://github.com/openshift/origin/blob/221f9a29d74a600fe30abd832a5c48810fc6d559/pkg/deploy/api/v1/types.go#L383
[2]
https://github.com/kubernetes/kubernetes/blob/2296108886a29db5cb7be73412eb562cdbf1cb74/pkg/apis/extensions/v1beta1/types.go#L268

On Thu, Jul 7, 2016 at 10:19 PM, Solly Ross <notifications@github.com>
wrote:

> The issue here is that having the annotation on the DC/RC/etc requires
> making the controller aware of the specific objects that it's dealing with
> (it needs a general client, etc). Additionally, it then either has to make
> an additional call besides the scale call to update the annotation (which
> means another place for things to get out of sync), or has to learn how to
> edit the scale of the individual object types (breaking the scale
> abstraction and making it less generic). It's not impossible to add the
> annotation, but it makes things more complicated.
>
> On Thu, Jul 7, 2016, 15:34 Ben Petersen <notifications@github.com> wrote:
>
> > I'm looking at this from the perspective of supporting idled deployments
> > in the web console (trello <https://trello.com/c/a5rIil2b>). I'd like to
> > avoid having to open another watch on Endpoints. If its possible to get
> the
> > annotation applied to the DCs directly that would definitely be
> beneficial.
> >
> > @DirectXMan12 <https://github.com/DirectXMan12> @smarterclayton
> > <https://github.com/smarterclayton> @jwforres
> > <https://github.com/jwforres> per some conversations we had earlier
> today.
> >
> > —
> > You are receiving this because you were mentioned.
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/openshift/origin/pull/9473#issuecomment-231183523>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/AAHTuczUv-TI_G2vy1SW7r7G4iVhTFZXks5qTVSngaJpZM4I7Lau
> >
>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9473#issuecomment-231195095>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADuFf15yhiVhT2p2We-xEB8PfBtxOeGDks5qTV9qgaJpZM4I7Lau>
> .
>

deads2k commented on pkg/unidling/controller/controller.go:
```diff
@@ -0,0 +1,323 @@
+		delete(targetEndpoints.Annotations, unidlingapi.IdledAtAnnotation)
+	} else {
+		var newAnnotationBytes []byte
+		newAnnotationBytes, err = json.Marshal(newAnnotationList)
```
Make sure you update to use proper conversions and serialization.  This ends up being part of the API.

deads2k commented on pkg/unidling/api/types.go:
```diff
@@ -0,0 +1,20 @@
+type RecordedScaleReference struct {
+	// TODO: update this to a CrossVersionObjectReference once we pull in a Kube version
+	// with an unversioned apis/autoscaling
+	kextapi.SubresourceReference `json:",inline"`
```
This is a known bad type, we shouldn't propagate it.  It uses kind instead of resource, it stuffs group information into an APIversion field, it suggests that auto-negotiation can't or won't happen.  You should make a subresource reference that works, group (required), version (optional, support negotiation), resource (this isn't a kind), name, subresource.

Also, looks like `CrossVersionObjectReference` has managed to reproduce the same flaws.

DirectXMan12 replied:
Feel free to correct me if I'm wrong here, but that suggestion seems to completely go against existing API design -- all the other reference types, AFAICT, refer to `Kind` instead of `Resource` and `APIVersion` in the slash-separated form (e.g. ObjectReference, OwnerReference, and group-versions are written as `APIVersion` in `TypeMeta` as well).  Furthermore, the scale client interfaces refer to `Kind` as well, so if we took a resource, we'd just have to convert it back to kind anyway (furthermore, I don't think we want to explicitly state that we're referring to the scale subresource here -- just an object that happens to have a scale subresource.  I'd much rather use `unversioned.CrossVersionObjectReference`, but that wasn't in the rebase I was working on).

While I'm not necessarily against the changes you're suggesting here, it seems to me like it would be confusing at best to have yet another object reference type (especially since CrossVersionObjectReference and SubresourceReference were specifically designed for referring to things that needed to be scaled) that seems like it would be completely different than all the other reference types, and would use conventions different than the rest of the API.

deads2k replied:
> Feel free to correct me if I'm wrong here, but that suggestion seems to completely go against existing API design -- all the other reference types, AFAICT, refer to Kind instead of Resource and APIVersion in the slash-separated form (e.g. ObjectReference, OwnerReference, and group-versions are written as APIVersion in TypeMeta as well).

Resource and kinds are different. Scale is a great example.  Scale is a kind, used by many resources.  We also have Binding which could logically be re-used.  If we accept that polymorphic endpoints take Kinds (they have to or the system doesn't work), then kind does not always map one to one to resource.  

Note that this system also failed for the origin and kube rbac types, where we wanted one Kind for multiple resources and people's bad assumptions on those one to one mappings (which don't even hold today) broke us.

`APIVersion` is used a slash delimited field, but in the case of a reference are you likely to care which version is used?  Almost never.  If both servers speak v2, use v2.  If they both speak v1, use it.  The only case where a distinction matters is when a resource changes meaning between versions.  This results in an APIVersion field that should almost never contain a version.  That's a pretty bad misnomer.



> Furthermore, the scale client interfaces refer to Kind as well, so if we took a resource, we'd just have to convert it back to kind anyway (furthermore, I don't think we want to explicitly state that we're referring to the scale subresource here -- just an object that happens to have a scale subresource. I'd much rather use

The current scale implementation is broken.  It takes a kind and assumes that a lowercase and pluralization will give it a resource.  Obviously that fails in the general case since the very first thing we did between versions was change Minion to Node so a Minion kind has a nodes resource.  There is a way to handle the mapping, but again, its a one to many, so you can end up failing.  You can see comments in the pulls establishing the current implementation point out these flaws.

tldr Kube did it wrong and has produced code that doesn't work right for the types they have in existence today.

deads2k replied:
Concretely, if you rely on `ScaleInterface.Update`, you are broken right here: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/scale.go#L66 .  While you are likely to use that API, you should avoid having bad data until just before your call.

DirectXMan12 replied:
Sure, that makes sense.  I don't disagree with what you're saying.   I'm just not overly a fan of doing different things across different parts of the API.  If we have a plan to replace uses of `CrossVersionObjectReference` with `ActuallyCorrectlyReferringToThingsObjectReference`, that's fine, but just having a different way to refer to things for this one piece here seems like it would be confusing, and wouldn't help to clarify the distinction between kind and resource.

deads2k replied:
> Sure, that makes sense. I don't disagree with what you're saying. I'm just not overly a fan of doing different things across different parts of the API.

If 100 people jump off a cliff....

This is API, but it's not directly user facing.  Someday this will be fixed upstream and HPA will have a half-working API.

DirectXMan12 replied:
🆗

deads2k replied:
> but just having a different way to refer to things for this one piece here seems like it would be confusing, and wouldn't help to clarify the distinction between kind and resource.

Spoke with @smarterclayton in person.  I'm not sure if this will make your day better or worse:
 1. Kind is ok as long as you enforce a single RESTMapping based on discovery for your server to locate the correct (and only) resource for a kind.
 1. APIVersion should be split into `Group` and `APIVersion`.  The apiversion is optional, but if it is specified, the group is ignored (this is backwards compatible).  The group is logically required, but empty for legacy groups.  If no apiversion is specified, the code should negotiate for the best available and compatible group version.
This logic is nasty, but we're describing a link, not a serialization format (resource, not kind).  Since group,resource,name represents the same thing across versions, the user wants to describe his intent for the link.  He doesn't want to upgrade his server and have his links break because he was forced to reference a version that eventually got sunset.


I tried and failed to convince him that the semantic for apiVersion was unnecessary for a new API, but if @liggitt wants to try.  Maybe we could even do without apiVersion entirely.

DirectXMan12 commented:
> There are other subresources (see DeploymentConfigRollback[1] in Origin or DeploymentRollback[2] in Kubernetes) that use an UpdatedAnnotations field which passes down those annotations to the object it updates. Scale could probably have something similar, no?

Sure, that seems somewhat reasonable, but we'd have to change/introduce a new version of the scale API.

smarterclayton commented on pkg/proxy/userspace/loadbalancer.go:
```diff
@@ -0,0 +1,34 @@
+/*
```
Can you open a PR upstream with this code and start the clock on review there?

smarterclayton replied:
Can you change the commit message so you link to the Kube PR on this?

DirectXMan12 replied:
done

smarterclayton commented on pkg/cmd/server/bootstrappolicy/policy.go:
```diff
@@ -465,6 +465,7 @@ func GetBootstrapClusterRoles() []authorizationapi.ClusterRole {
 			},
 			Rules: []authorizationapi.PolicyRule{
 				authorizationapi.NewRule("list", "watch").Groups(kapiGroup).Resources("endpoints").RuleOrDie(),
+				authorizationapi.NewRule("list", "watch").Groups(kapiGroup).Resources("services").RuleOrDie(),
```
With the annotation moved to endpoints is this still necessary?

DirectXMan12 replied:
The router has to be able to get the Service IP, for when the route is idled.  The router doesn't actually send the unidling event -- it just forwards to the service proxy when it sees something that's idled.

smarterclayton commented on pkg/router/template/plugin.go:
```diff
@@ -253,11 +254,41 @@ func peerEndpointsKey(namespacedName ktypes.NamespacedName) string {
+			return []Endpoint{}
+		}
+
+		svcSubset := kapi.EndpointSubset{
```
This assumes that clusterIP is set, which does not have to be the case.  For instance, Headless services have the field set to "None", so you have to handle and deal with that here, and probably also guard it in "idle" (it's not clear to me that we can safely idle something that is headless with the current code).

DirectXMan12 replied:
Yeah, there's not a good way to do unidling with headless services.  I'll make sure we handle that case gracefully.

DirectXMan12 commented:
> 1. Kind is ok as long as you enforce a single RESTMapping based on discovery for your server to locate the correct (and only) resource for a kind.

This is something that would have to be changed in the ScaleClient, etc code, no?

> 2. APIVersion should be split into Group and APIVersion. The apiversion is optional, but if it is specified, the group is ignored (this is backwards compatible). The group is logically required, but empty for legacy groups. If no apiversion is specified, the code should negotiate for the best available and compatible group version.

So... we need a new type with split `group` and `apiVersion` fields, but that's about it for the time being?  The current ScaleClient code just complete ignores GVK (which is why we have the `DelegatingScaleNamespacer` nonsense).

> Spoke with @smarterclayton in person. I'm not sure if this will make your day better or worse

Unless it's being proposed that we fix the ScaleClient code in this PR (which makes my day worse), I don't entirely think this changes my day much (I just need to add the new type, which shouldn't be too bad) :neutral_face:

deads2k commented:
> Unless it's being proposed that we fix the ScaleClient code in this PR (which makes my day worse)

Doesn't have to be in this PR, but it does need doing or we aren't going to have much luck trying to make the HPA scale generically scalabe things.

DirectXMan12 commented:
Ok, I've added in some extended tests which test idling and unidling, from both the TCP and UDP perspective.  While writing the UDP test, @knobunc and I discovered an edge case with UDP involving continuing to use a the same socket and UDP conntrack.  We're working on a fix ATM.

I've also added in a new type called `CrossGroupObjectReference` with need split fields.  Currently the fields are ignored by the controller, since the scale API doesn't use the 😞 .

I still have to fix the ClusterIP assumption, but I think I've gotten all the other changes requested.

DirectXMan12 commented:
ok, router should skip headless services now.  @knobunc is whipping up a code to solve the conntrack UDP problem right now, sot hat should be in shortly.

benjaminapetersen commented:
Refreshing the discussion via [this comment](https://github.com/openshift/origin/pull/9473#issuecomment-231183523) [this comment](https://github.com/openshift/origin/pull/9473#issuecomment-231399754).  Still curious if we can get annotations on the DC/RC as part of this.

DirectXMan12 commented:
@smarterclayton I've updated the PR to scale and update DeploymentConfigs and ReplicationControllers by using their own clients, instead of via Scales.  This allows us to add and remove the idled-at annotation on DCs and RCs, so @benjaminapetersen should be able to use that to see if a DC or RC is idled.

@benjaminapetersen does that work ok for you?
@smarterclayton besides any review comments from that, is there anything else you want from this PR?

benjaminapetersen commented:
@DirectXMan12 yup, I believe that covers what I need. Thanks!

smarterclayton commented:
Before we merge this, can you open the PRs to kube that contain

1. Updated proposal for idling
2. The refactored proxy (no idler specific)

smarterclayton commented on pkg/router/controller/factory/factory.go:
```diff
@@ -87,6 +95,15 @@ func (factory *RouterControllerFactory) Create(plugin router.Plugin) *controller
 			return routeEventQueue.ListConsumed()
 		},
+		LookupService: func(endpoints *kapi.Endpoints) (*kapi.Service, error) {
+			if rawSvc, ok, err := svcStore.Get(endpoints); err != nil {
```
Style, don't do elses when returning.

smarterclayton commented on pkg/router/controller/controller.go:
```diff
@@ -111,6 +112,12 @@ func (c *RouterController) HandleEndpoints() {
 	}
 
+	// NB: svc might be nil here
+	svc, err := c.LookupService(endpoints)
```
I don't like this - I think the lookup for service should be done by the specific handler, not by passing this down.  When we switch to informer patterns in the router, there will be no way to pass this down, so let's not start now.

smarterclayton commented on pkg/router/template/plugin.go:
```diff
@@ -155,7 +157,7 @@ func (p *TemplatePlugin) HandleEndpoints(eventType watch.EventType, endpoints *k
 	case watch.Added, watch.Modified:
 		glog.V(4).Infof("Modifying endpoints for %s", key)
-		routerEndpoints := createRouterEndpoints(endpoints, !p.IncludeUDP)
+		routerEndpoints := createRouterEndpoints(endpoints, !p.IncludeUDP, service)
```
As mentioned above I'd prefer TemplatePlugin to have LookupService and pass that method in to the createRouterEndpoints (which then only does the lookup on changes)

DirectXMan12 replied:
just to clarify, are you suggesting here that the TemplatePlugin just maintain it's own service listwatch cache (or whatever), and just use that (so, pass a client to the TemplatePlugin constructor, etc)?

smarterclayton replied:
Actually accept an interface which is "GetService" that is backed by your cache.  That's an interface the template router gets passed in to it, that happens to be backed by the cache (i'm actually tempted to say we should just be fetching the service as necessary, but either way the abstraction is that template router is calling an opaque interface to get the service info, and deals with the info not being present correctly, instead of getting a service passed down).

DirectXMan12 replied:
ack.  Just wanted to make sure I was heading down the right path ;-)

smarterclayton commented on pkg/router/controller/factory/factory.go:
```diff
@@ -60,11 +62,17 @@ func (factory *RouterControllerFactory) Create(plugin router.Plugin) *controller
 		// we do not scope endpoints by labels or fields because the route labels != endpoints labels
 	}, &kapi.Endpoints{}, endpointsEventQueue, factory.ResyncInterval).Run()
 
+	svcStore := cache.NewStore(cache.MetaNamespaceKeyFunc)
```
You have to handle permission denied here - routers will be upgraded and may not have access to services until an admin upgrades their permissions.  So you need to test against a server without permission and ensure that the cache is loaded, warns the admin (via a log message) and then works correctly.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+	idleLong = `
+Idle scalable resources.
+
+This command idles the provides list of resources by scaling scalable resources (e.g. RCs and
```
Typo.  Don't use "scalables", sounds awkward.  Describe what idling is first, then describe what can be idled, then describe what happens on idling.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+DCs) down to zero replicas, and then marking associated services so that that the scalables
+will be "woken up" when traffic occurs on those services.
+
+Only DCs and RCs with associated services will be idled -- services that only point to pods,
```
What about pet sets?

Also, don't use DC and RC in descriptions, say "deployment config" or "replication controller".

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+	}
+
+	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "If true, only print the annotations that would be written, without annotating or idling the relevant objects")
+	cmd.Flags().StringVarP(&inputFile, "filename", "f", "", "file containing list of services whose scalables will be idled")
```
Don't use "scalables".

This command needs to support the normal resource builder label selectors and name inputs, in addition to `-f -`.  So at minimum `--all`, `--all-namespaces`, `-l`, and resource type name (use the form that defaults to services, so someone can do `oc idle foo` as well as `oc idle svc/foo`.

smarterclayton replied:
(I saw below you default to endpoints, that's fine).

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+
+	err = b.Do().Visit(func(info *resource.Info, err error) error {
+		endpoints := info.Object.(*api.Endpoints)
+		idlableRefs, err := findIdlablesForEndpoints(endpoints, decoder, getPod, getController)
```
Don't use `idlable`, it's not a real word.  `findReferencesToIdleForEndpoints`

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+
+// RunIdle runs the idling command logic, taking a list of resources or services in a file, scaling the associated
+// scalables to zero, and annotating the associated endpoints objects with the scalables to unidle when they receive traffic.
+func RunIdle(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, filename string, loadByService, dryRun bool) error {
```
You need to be using the `options` pattern.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+
+		if !dryRun {
+			if len(info.scalables) == 0 {
+				fmt.Fprintf(out, "No scalables marked as idled for service %s/%s, not marking as idled...\n", namespace, serviceName)
```
You need to follow our consistent output scheme - all errors are prefixed with `error: ` or `warning: ` and are printed to stdout.  If `-o name` is printed you need to be printing `resource/name` (we can defer that for now potentially, but leave a todo).  All other output goes to stdout.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+				continue
+			}
+			if _, err := patchObj(info.obj, metadata, oldData, mapping, f); err != nil {
+				fmt.Fprintf(out, "Unable to mark service %s/%s as idled: %v", namespace, serviceName, err)
```
Every error needs to be counted and at the end and if count > 0 at the end you should return exit code 1.

smarterclayton replied:
You should also print a summary when `-o name` is not set (take the structure into account).

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,557 @@
+
+		fmt.Fprintf(out, "Marked service %s/%s to unidle scalables:\n", namespace, serviceName)
+		for _, scaleRef := range refsWithScale {
+			fmt.Fprintf(out, "    %s %s/%s (unidle to %v replicas)\n", scaleRef.Kind, namespace, scaleRef.Name, scaleRef.Replicas)
```
Don't do multiline prints like this - have each line be self contained.

smarterclayton commented on pkg/cmd/cli/cmd/idle_test.go:
```diff
@@ -0,0 +1,316 @@
+package cmd
```
This needs test/cmd tests.  You should create synthetic endpoints and verify idle works on them.

smarterclayton replied:
Specifically testing the CLI command and its interactions, not necessarily the whole feature (which is what extended tests are for)

DirectXMan12 commented:
> The refactored proxy

Sure.  I'll submit a PR tomorrow (I'll be offline this afternoon).

> Updated proposal for idling

I'll write something up, but I suspect upstream will not want exactly what we have here (e.g. events will probably be problematic, and I suspect a few people, at least, will want an HPA-like object instead of just a command).

DirectXMan12 commented:
Alright, I've fixed up the CLI to use the options pattern, better nomenclature, and accepting label selectors.  I'm just finishing up some test/cmd tests.

DirectXMan12 commented:
✔️ test/cmd tests added

DirectXMan12 commented:
The refactored proxy: kubernetes/kubernetes#29465

DirectXMan12 commented:
Idling proposal: kubernetes/kubernetes#29471

smarterclayton commented on pkg/sdn/plugin/endpoints.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+func (oc *OsdnNode) EndpointStartNode() error {
+	go utilwait.Forever(oc.watchEndpoints, 0)
```
Does this mean now that the node is watching endpoints multiple times again, from multiple places?  When the CNI changes land, will this start watching endpoints from two processes on the same node?

knobunc replied:
I'll coordinate with @dcbw to make sure we don't.  But the CNI work is not landing until 3.4.

smarterclayton commented on pkg/sdn/plugin/bin/openshift-sdn-ovs:
```diff
@@ -9,11 +9,14 @@ tenant_id=$3
 ingress_bw=$4
```
Do we need this commit in this PR for the core idling work to land?  If not I'd like it split out given the perf concern raised below.

DirectXMan12 commented:
as per a conversation with @knobunc @smarterclayton, we're splitting the OVS rules commit out into a separate PR.  I think this should be good to go now.

smarterclayton commented on pkg/cmd/server/kubernetes/node.go:
```diff
@@ -421,14 +423,29 @@ func (c *NodeConfig) RunProxy() {
 	serviceConfig := pconfig.NewServiceConfig()
-	serviceConfig.RegisterHandler(proxier)
+
+	unidlingLoadBalancer := ouserspace.NewLoadBalancerRR()
```
Can I disable the unidler if I want to?  I think we need a way to be able to turn this off if necessary, which means a config flag or ProxyConfig.Mode or something similar.

DirectXMan12 replied:
Is it acceptable to add additional proxy modes in an OpenShift specific place?  If so, I could just add a "idling-iptables" and "idling-userspace" proxy modes.  Otherwise, I'll just add a flag.

smarterclayton replied:
If someone wanted to change between iptables and userspace, would they necessarily want to decide "idling on/off" at the same time?  That's an argument for a flag in the config.

DirectXMan12 replied:
Fair point.  I'll add a separate option.

smarterclayton commented on pkg/unidling/api/types.go:
```diff
@@ -0,0 +1,7 @@
+package api
+
+const (
+	IdledAtAnnotation      = "idling.alpha.openshift.io/idled-at"
```
Needs Godoc on what they do

DirectXMan12 replied:
✔️

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,630 @@
+	}
+
+	cmd.Flags().BoolVar(&o.dryRun, "dry-run", false, "If true, only print the annotations that would be written, without annotating or idling the relevant objects")
+	cmd.Flags().StringVarP(&o.filename, "filename", "f", o.filename, "file containing list of services whose scalable resources to idle")
```
I missed this before, but we shouldn't be using `filename` for this since it's not the same as all the other uses of `-f`.  We should call this something different, like `resource-names-file` (it doesn't need a short cut flag).

DirectXMan12 replied:
✔️

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,630 @@
+}
+
+// patchObj patches calculates a patch between the given new object and the existing marshaled object
+func patchObj(obj runtime.Object, metadata meta.Object, oldData []byte, mapping *meta.RESTMapping, f *clientcmd.Factory) (runtime.Object, error) {
```
There should be a helper for this - is that not usable for you?

DirectXMan12 replied:
IIRC, there was some reason it wasn't quite usable when I originally wrote this.  I'll take a look now

DirectXMan12 replied:
Actually, I think this was just a wrapper around a couple lines that are just inline elsewhere in the code base, because I needed to do the same thing multiple times at one point.  IMO, it's a bit clearer like this, even though it's just used one place.

smarterclayton commented on pkg/cmd/cli/cmd/idle.go:
```diff
@@ -0,0 +1,630 @@
+	for scaleRef, svcName := range byScalable {
+		obj, scale, err := scaleAnnotater.GetObjectWithScale(svcName.Namespace, scaleRef)
+		if err != nil {
+			fmt.Fprintf(o.errOut, "error: unable to get scale for %s %s/%s, not marking that scalable as idled...\n", scaleRef.Kind, svcName.Namespace, scaleRef.Name)
```
No ellipsis here.

DirectXMan12 replied:
✔️

smarterclayton commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -73,6 +73,9 @@ type NodeConfig struct {
 	IPTablesSyncPeriod string `json:"iptablesSyncPeriod"`
 
+	// DisableUnidling controls whether or not the hybrid unidling proxy will be set up
+	DisableUnidling bool `json:"enableUnidling"`
```
We have positive flags that default if not set.

```
EnableUnidling *bool `json:"enableUnidling"`
```

And add a defaulter in v1 to default nil to true.  The internal type does not need to be a bool pointer, just a bool.

DirectXMan12 replied:
ack.

ncdc commented:
Need to fix `error: converting (api.NodeConfig) to (v1.NodeConfig): type names don't match (bool, ), and no conversion 'func (bool, *bool) error' registered.`

DirectXMan12 commented:
this should fix the issue -- @smarterclayton pointed me to a conversion that needed to be added to the config API scheme

DirectXMan12 commented:
[test]

smarterclayton commented on pkg/cmd/server/kubernetes/node.go:
```diff
@@ -421,14 +423,31 @@ func (c *NodeConfig) RunProxy() {
+		unidlingLoadBalancer := ouserspace.NewLoadBalancerRR()
+		unidlingUserspaceProxy, err := ouserspace.NewProxier(unidlingLoadBalancer, bindAddr, iptInterface, execer, *portRange, c.ProxyConfig.IPTablesSyncPeriod.Duration, c.ProxyConfig.UDPIdleTimeout.Duration)
+		if err != nil {
+			glog.Warningf("WARNING: Could not initialize Kubernetes Proxy. You must run this process as root to use the service proxy: %v", err)
```
Actually, this should no longer be a warning (the other code was changed to be fatal).  Be consistent with the other bits.

DirectXMan12 commented:
@smarterclayton I think we have to have the internal version be a pointer as well.  Otherwise, when we generate config starting with an internal version (e.g. with `openshift start --write-config`, I believe), we'll get `enableUnidling: false`, since we're starting with a normal bool (which defaults to false).

(I haven't made the change to use a pointer in the internal version yet though).

smarterclayton commented:
There's code that handles this by calling explicit defaults when writing out config.  The protobuf stuff uses it.

DirectXMan12 commented:
Are you talking about this: https://github.com/openshift/origin/blob/master/pkg/cmd/server/start/node_args.go#L198 ?

I've tested, and it gets the wrong value when generating.  Do I just need to explicitly set the value above when we create the config?  Otherwise, it looks like the `setDefaults` function wont help here, because we'll go `internal.UnidlingEnabled = false --[convert to v1]--> v1.UnidlingEnabled = &false --[convert to internal]--> [v1.UnidlingEnabled != nil, so don't change it] internal.UnidlingEnabled = false`

smarterclayton commented:
Yes, explicitly set value.  Also be aware there are two places this
config gets generated so you have to change both.

DirectXMan12 commented:
alright, done, time to run the [test]

smarterclayton commented:
One real bug, one integration flake.  @deads2k
TestAuthorizationResourceReview - known?

ncdc commented:
[test] #10008

DirectXMan12 commented:
cc @deads2k to double check my ResourceAccessReview test pretty printing changes

deads2k commented on test/integration/authorization_test.go:
```diff
@@ -27,6 +28,63 @@ import (
+	return fmt.Sprintf(nsStr + usersStr + groupsStr)
+}
+
 func TestClusterReaderCoverage(t *testing.T) {
```
test case and printing changes look fine.

DirectXMan12 commented:
@openshift-bot appears to have had a problem building images.  :weary:

ncdc commented:
[test]

On Friday, August 5, 2016, Solly Ross <notifications@github.com> wrote:

> @openshift-bot <https://github.com/openshift-bot> appears to have had a
> problem building images. 😩
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9473#issuecomment-237996255>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYhXwla440-ft_80YkOMVT8eotsD5ks5qc9p5gaJpZM4I7Lau>
> .
>

smarterclayton commented:
[test]

smarterclayton commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -771,6 +774,49 @@ func init() {
+					Verbs:     sets.NewString("get", "update"),
+					Resources: sets.NewString("endpoints"),
+				},
+				// these should be removed once we can set the last-scale-reason field via the scale subresource
```
Where are these used?

Also, you should add petset.

DirectXMan12 commented:
@smarterclayton added `/scale` to permissions for `Deployment` and `ReplicaSet`.  Did you want the the specialized scaling for those as well?  Also, it looks like PetSet doesn't have a scale subresource yet.  Does PetSet support scaling w/o any wrinkles?  kubernetes/contrib#1295 seems to imply scaling doesn't quite work yet...

smarterclayton commented:
It's fine to leave as is.

On Tue, Aug 9, 2016 at 10:10 AM, Solly Ross <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> added /scale to
> permissions for Deployment and ReplicaSet. Did you want the the
> specialized scaling for those as well? Also, it looks like PetSet doesn't
> have a scale subresource yet. Does PetSet support scaling w/o any wrinkles?
> kubernetes/contrib#1295 <https://github.com/kubernetes/contrib/pull/1295>
> seems to imply scaling doesn't quite work yet...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9473#issuecomment-238564999>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p00sQwu2-U54qno-Gri8kHNh3SeSks5qeIpegaJpZM4I7Lau>
> .
>

smarterclayton commented on pkg/cmd/server/origin/run_components.go:
```diff
@@ -540,3 +543,15 @@ func (c *MasterConfig) RunIngressIPController(client *kclient.Client) {
+	scaleNamespacer := osclient.NewDelegatingScaleNamespacer(oc, kc)
+	coreClient := clientadapter.FromUnversionedClient(kc).Core()
+	dcCoreClient := deployclient.New(oc.RESTClient)
+	cont := unidlingcontroller.NewUnidlingController(scaleNamespacer, coreClient, coreClient, dcCoreClient, coreClient, resyncPeriod)
```
Open a follow up item to use a shared informer for endpoints
